### DISCUSSION TITLE
Internal Server Error and other issues

### DISCUSSION THREAD

**Discussion by privileger** [Azure-Samples/chat-with-your-data-solution-accelerator#1303] [General]
Hi everyone,

I deployed this repo a couple of months ago and used it without any issues. Today, I deployed it again in a demo environment, but the front-end application is not running properly, showing an 'Internal Server Error.' The backend application is running; however, after uploading the document for indexing, it is throwing multiple errors.

The template was deployed without modifications. I only added the Japanese language "ja-JP" in the list of supported languages during the deployment. 

Please advise how to resolve this issue.

![000](https://github.com/user-attachments/assets/d41cb619-5fc1-4883-9640-59697f919fad)
<img width="335" alt="001" src="https://github.com/user-attachments/assets/371b4d0d-4b5d-419e-8a96-11ece4385f3e">
<img width="950" alt="002" src="https://github.com/user-attachments/assets/11a8951b-1e8f-46a7-91de-8fe5e6666245">

**Comment by Roopan-Microsoft** [comment]
Hey @privileger, As mentioned in the issue. Please try to redeploy as this fix has been merged as part of #1305.

Please raise an issue in case you face any blocker or issue.

Happy to help!

**Reply by privileger** [reply]
@Roopan-Microsoft 
Thanks for the update. I just re-deployed this repo, and the reported issue is gone. However, the front end is still acting strange because I am receiving an error: "An error occurred. Please try again. If the problem persists, please contact the site administrator." After clicking the OK button, the LLM no longer returns answers based on the RAG's content. Please advise.

![error20240916](https://github.com/user-attachments/assets/2cfe438d-758c-464d-9641-0ce10cda559e)

**Reply by Roopan-Microsoft** [reply]
Hey @privileger, It looks strange. 

Could you please investigate further using Application Insights to check for any exceptions or additional details? I believe there might be relevant information that could help us diagnose the issue more precisely. 

Also please let us know if any configuration change happend, so that we can try to repro the same from our end.

<img width="939" alt="image" src="https://github.com/user-attachments/assets/52679dbf-9d7e-47e4-827c-6056d10b5e71">


Looking forward to your findings.

**Comment by privileger** [comment]
@Roopan-Microsoft 
Here you go.
![error20240916-2](https://github.com/user-attachments/assets/aec9a73c-d466-437f-8051-063e9a4eebcf)

**Comment by privileger** [comment]
@Roopan-Microsoft 
![error20240916-3](https://github.com/user-attachments/assets/8dc8b0ce-7010-4eda-a9c3-639aac65d104)
![error20240916-4](https://github.com/user-attachments/assets/8d4eaf74-d274-4f31-9bbd-8fc4024ba070)

Some additional screenshots.

**Reply by Prasanjeet-Microsoft** [reply]
@privileger In the last screenshot you provided, could you please click on the 500 response code and provide additional details? I have provided an example below for your reference.

**Click on 500 response code:**

<img width="878" alt="Screenshot 2024-09-17 014114" src="https://github.com/user-attachments/assets/3813191c-dc7a-4c94-a8aa-8bc932e1e4a3">


**Click on suggested view details:**

<img width="831" alt="Screenshot 2024-09-17 014228" src="https://github.com/user-attachments/assets/7189b4e5-a9d4-464f-a6fa-1e36f8b37247">


**Click on EXCEPTION:**

<img width="932" alt="Screenshot 2024-09-17 014933" src="https://github.com/user-attachments/assets/83d11aec-bea6-47e8-b7e1-81765b365d0e">

**Reply by privileger** [reply]
@Roopan-Microsoft Here you go. Hope it helps.
![error20240917-1](https://github.com/user-attachments/assets/d884b3f6-a217-4225-9815-59e5cc6dce3c)
![error20240917-2](https://github.com/user-attachments/assets/ba2f8818-fa68-4520-8e61-684a00a66d54)
![error20240917-3](https://github.com/user-attachments/assets/32690a22-72dc-42b7-89c6-ee7b49bb97c9)
![error20240917-4](https://github.com/user-attachments/assets/78474b32-57dd-422a-9580-278100e4b81e)
![error20240917-5](https://github.com/user-attachments/assets/ac95e7f9-5ff7-47de-9e52-ade68a93b366)
![error20240917-6](https://github.com/user-attachments/assets/0f60a7f0-ddea-43df-aef6-e9c43565a3fe)
![error20240917-7](https://github.com/user-attachments/assets/f2f637c4-5323-4933-9747-c1568b3f5db7)

**Reply by Roopan-Microsoft** [reply]
Hey @privileger - Thanks for more details, we are looking into this. Will keep you posted.

**Reply by Roopan-Microsoft** [reply]
Hey @privileger,

We see multiple issues, Let me explain one by one,

**Issue 1:** _( Open AI Content Management Policy)_
This issue is due to the question which may be harmful content or anything sensitive. Open AI will filter those and will respond as 400. Which is seems to be expected. (App should work for normal queries)
Let us know the question which you asked. we will verify from our end.
<img width="755" alt="image" src="https://github.com/user-attachments/assets/60829def-8f69-40f2-81b9-b3ff57ba8944">

**Issue 2:** _(No Index found)_
Based on the exceptions, we see there is missing index. Please check in your Search Service resource.
Serach Service -> Index (Check if you have any index called _conversations_)
<img width="750" alt="image" src="https://github.com/user-attachments/assets/8b25594d-07f9-4c0c-825b-4bce3ea0c4a1">

Please let us know your observations.

Also please let us know if your issue exists still or on specific scenarios.

Happy to Help!

**Comment by Prasanjeet-Microsoft** [comment]
@privileger Could you please provide us with an update on the above comments?

FYI @Roopan-Microsoft

**Reply by privileger** [reply]
@Roopan-Microsoft Hello, and thank you for the follow-up. I deployed a solution today in the Japan East region, and it worked for a while before the front end generated the same error. Please refer to the screenshots below. I’ve observed that the chatbot's functionality seems to depend on the context.

During deployment, I only modified one parameter, removing all languages except en-US and adding ja-JP.

<img width="382" alt="スクリーンショット 2024-09-24 132346" src="https://github.com/user-attachments/assets/b10042f6-d1f3-43f8-97ca-ab65bd543eb3">

I then uploaded a 129-page Word document written in Japanese and attempted to query the LLM based on its contents. The document included text, tables, and images. While it worked for some prompts, it eventually resulted in the same error as before.

Please refer to the screenshot below.

<img width="938" alt="スクリーンショット 2024-09-24 144836" src="https://github.com/user-attachments/assets/4752e67c-b8a0-4a60-8083-751b13f6f9da">

**Comment by privileger** [comment]
@Roopan-Microsoft 
Please find below some additional screenshots.
![01](https://github.com/user-attachments/assets/119aa3d8-f7a5-4637-8cd0-221b98537b31)
![02](https://github.com/user-attachments/assets/eb0f19d0-f84c-440c-a8f2-596f396fe4ea)
![2](https://github.com/user-attachments/assets/73470c34-ac95-43d5-9304-8bb151339203)
![03](https://github.com/user-attachments/assets/3ddce313-b3d1-4ecd-b36f-92cfa2fdd8d0)
![3](https://github.com/user-attachments/assets/161fa77a-dc3e-42ee-8d20-7bbb2fdc7ee1)
![04](https://github.com/user-attachments/assets/5ad91467-58f2-4c81-afc4-2a3d1549b12d)
![4](https://github.com/user-attachments/assets/846e184d-f6d1-44e1-bc72-dca900d4f70d)
![05](https://github.com/user-attachments/assets/32512c31-cf8e-43ee-84d4-b6a3cabccb03)
![06](https://github.com/user-attachments/assets/105094b1-c247-4f26-bb76-e05ffd9db5e0)
![07](https://github.com/user-attachments/assets/94edb050-2417-4fae-b0bd-4b8b064158dc)
![スクリーンショット 2024-09-25 13323](https://github.com/user-attachments/assets/db037277-4aac-40b7-9469-17d7dc412f03)
![スクリーンショット 2024-09-25 133139](https://github.com/user-attachments/assets/daabb836-b02b-43c6-8d25-3a138dd6258f)
![スクリーンショット 2024-09-25 133251](https://github.com/user-attachments/assets/32908719-7499-41a1-9715-087a96a107b0)
![スクリーンショット 2024-09-25 133323](https://github.com/user-attachments/assets/a3e85321-0106-4b10-bebc-b7df19776592)
![スクリーンショット 2024-09-25 133345](https://github.com/user-attachments/assets/9d0b279f-a35c-4745-8ce6-ac9a28a6219a)
![スクリーンショット 2024-09-25 133410](https://github.com/user-attachments/assets/0531cc68-6d9b-4456-b75f-0db490a491c2)
![スクリーンショット 2024-09-25 133428](https://github.com/user-attachments/assets/c455df8f-440d-4405-904d-8d91df98a878)
![スクリーンショット 2024-09-25 133457](https://github.com/user-attachments/assets/577b7c5e-d9b6-457f-a258-4c7376182402)
![スクリーンショット 2024-09-25 133514](https://github.com/user-attachments/assets/b1c97102-1107-4c88-b7be-de2462ea7847)
![スクリーンショット 2024-09-25 133538](https://github.com/user-attachments/assets/bc314a7f-d904-4e50-a78c-99ae6bf58424)
![スクリーンショット 2024-09-25 133559](https://github.com/user-attachments/assets/1a9a6475-13e4-435e-b293-aba362164651)
![スクリーンショット 2024-09-25 133619](https://github.com/user-attachments/assets/a50bc0a0-599d-4873-b5c2-5fd75dd940bd)
![スクリーンショット 2024-09-25 133637](https://github.com/user-attachments/assets/a342440a-12c8-44bd-92bb-5ab7c6b8fe84)
![スクリーンショット 2024-09-25 133656](https://github.com/user-attachments/assets/eced891a-dc6f-4bb3-ae07-ee28edd3b032)
![スクリーンショット 2024-09-25 1332521](https://github.com/user-attachments/assets/c68595ac-e8ff-43f9-9c37-cf6e329c4325)

**Comment by Prasanjeet-Microsoft** [comment]
Hello @privileger , After reviewing the screenshots, it looks like you are experiencing two types of issues: one related to **content management policy** and the other concerning **authorization for machine learning services**.

**1st Issue: Content Management Policy**

- As mentioned in the previous comment, this issue arises from the content filtering system, which identifies and takes action on certain categories of potentially harmful content in both input prompts and output completions. This behavior appears to be expected.

- Therefore, we recommend asking different questions or 
- Try to deploy this accelerator using the **gpt-4o (2024-05-13) GlobalStandard model**, which offers superior performance in non-English languages, if you intend to deploy only in the Japan East region and ask questions in japanese language. 
- Alternatively, you could also deploy this accelerator in a different region using the **gpt-4o (2024-05-13) model** and **gpt-4o (2024-08-06) model** to achieve better performance in non-English languages.

Please find a few links below for your reference:
https://learn.microsoft.com/en-us/azure/ai-services/openai/concepts/content-filter?tabs=warning%2Cuser-prompt%2Cpython-new
https://learn.microsoft.com/en-us/azure/cognitive-services/openai/concepts/models#model-summary-table-and-region-availability
https://learn.microsoft.com/en-us/azure/ai-services/openai/concepts/models#global-standard-model-availability
https://learn.microsoft.com/en-us/azure/ai-services/openai/concepts/models#fine-tuning-models

**2nd Issue: Authorization for Machine Learning Services**

- It seems the service principal that you are using doesn't have rights within that tenant.

- Tenants have subscriptions and service principals belong to tenants. Azure resource manager also exposes role based authorization for a given principal, which would give it rights on Azure resources. It appears the service principal doesn't have rights to read from that subscription.
- Go to portal and find your subscription, click on Access Control (IAM) and then click on Add role assignment with correspond service principal which you use to acquire token.

- After you have given successful permission, refresh and try again.

**Note: If you dont have service principal then you can follow below steps**

**To check access in an Azure subscription and assign the Reader or Contributor role, follow these steps:**

**Step 1: Sign in to the Azure Portal**

1. Go to the [Azure Portal](https://portal.azure.com/).
2. Sign in with your Azure account credentials.

**Step 2: Check Access in the Subscription**

1. In the left-hand navigation pane, select Subscriptions.
2. Choose the subscription you want to check.
3. In the subscription menu, click on Access control (IAM).
4. On the Check access tab, click the View my access button.
5. An assignments pane appears that lists your access at this scope and inherited to this scope. If you don't have reader or contributor role then follow further steps.

**Step 3: Assign the Reader or Contributor Role**

1. In the Access control (IAM) section, click on the + Add button.
2. Select Add role assignment.
3. On the Role tab, select a role that you want to use. You can search for a role by name or by description. (Select either Reader or Contributor, depending on the level of access you want to assign)
4. Click Next
5. On the Members tab, select User, group, or service principal to assign the selected role to one or more Microsoft Entra users, groups, or service principals (applications).
6. Click Select members.
7. Find and select the users, groups, or service principals. (You can type in the Select box to search the directory for display name or email address.)
8. Use the Select option to find and select the user (your name) or group you want to assign the role to.
9. Click Review + assign to assign the role
10. After a few moments, the security principal is assigned the role at the selected scope

**Step 4: Confirm the Assignment**

1. After saving, you should see the new role assignment listed in the role assignments section.

**Notes**

- Make sure you have the necessary permissions to assign roles in the subscription (typically, you need to be an Owner or have the User Access Administrator role).
- Your updates might take a while to be processed and reflected in the portal.

Please find below links for your reference:
https://learn.microsoft.com/en-us/azure/role-based-access-control/check-access#step-2-check-access-for-a-user
https://learn.microsoft.com/en-us/azure/role-based-access-control/role-assignments-portal
https://learn.microsoft.com/en-us/azure/role-based-access-control/quickstart-assign-role-user-portal

**Comment by Prasanjeet-Microsoft** [comment]
@privileger Could you please let us know if you’re able to proceed ahead without any issues? If so, can we close this issue?

**Reply by Prasanjeet-Microsoft** [reply]
> Could you please let us know if you’re able to proceed ahead without any issues? If so, can we close this issue?

Hello @privileger, just checking in to see if you’re able to proceed without any issues. If everything looks good, can we go ahead and close this issue?

Thank you!

**Comment by privileger** [comment]
@Prasanjeet-Microsoft I need a bit more time to redeploy the solution and check everything.

All Incan say for now that the document I fed to Azure Open AI is related to the health and safety, so I am not sure why the content was blocked by content management policy. Can I get more details in this? 

Also, what parameters should be changed in order to activate 4o model?

**Comment by Prasanjeet-Microsoft** [comment]
@privileger Do you prefer to deploy in the Japan East region or is any region acceptable?

**Reply by privileger** [reply]
@Prasanjeet-Microsoft For the testing purposes, it can be deployed in any region, however, the production environment must be deployed in Japan East.

**Comment by Prasanjeet-Microsoft** [comment]
Hello @privileger, If you want to activate the gpt-4o model, you need to update the default values of the parameters to the new values listed below during the new deployment:

**Azure Open AI Model : gpt-4o
Azure Open AI Model Name : gpt-4o
Azure Open AI Model Version : 2024-05-13**

Please find screenshot for your reference:

<img width="950" alt="image" src="https://github.com/user-attachments/assets/361a8880-f0a9-49c9-9a39-cc9aefb481fc">


Note: **Please refer below link for Model summary table and region availability before doing the deployment with gpt-4o model**

https://learn.microsoft.com/en-us/azure/cognitive-services/openai/concepts/models#model-summary-table-and-region-availability

**Comment by privileger** [comment]
@Prasanjeet-Microsoft  Hello. Thank you for resolving this issue. I will close a ticket for now.
